platform/darwin/src/MGLFeature.mm

```diff
@@ -51,6 +58,56 @@ - (NSString *)description
             self.attributes.count ? self.attributes : @"none"];
 }
 
+- (id<MGLPointCluster>)cluster {
+    return self.cluster ? self : nil;
+}
+
+#pragma mark - MGLPointCluster
+
+- (BOOL)isCluster {
+    NSNumber *isCluster = MGL_OBJC_DYNAMIC_CAST(self.attributes[MGLPointClusterBoolKey], NSNumber);
+    return [isCluster boolValue];
+}
+
+- (uint32_t)clusterId {
```

Comment:
Can you go into the decision to keep `uint32_t` as the public type for this? I’m concerned about Swift bridging and Obj-C ergonomics.

Reply:
The main deciding factor would be whether a type is capable of representing the maximum cluster ID.

`uint32_t` bridges to Swift as `Int32`, which isn’t the end of the world but also fairly unusual. Normally, the only APIs that use that type are ones that deal with low-level devices such as file handles.

`mbgl::AnnotationID` is similarly `uint32_t`, but `MGLAnnotationTag` is `NSUInteger` instead, for familiarity’s sake. mbgl treats GeoJSON’s number type as `uint64_t` or `int64_t` depending, but the iOS/macOS map SDK uses (unsigned) `long long`, which bridges to Swift as (U)`Int64` regardless.

Reply:
How about we have a typedef? Something like `MGLClusterIdentifier`?

Reply:
Note that a typedef has implications for Swift code: you have to cast even between MGLClusterIdentifier and the raw type. Perhaps that extra type safety is warranted; perhaps not. But the more important question @friedbunny raises is about the raw data type: what are the minimum and maximum values that this identifier can conceivably be set to without underflowing or overflowing?

Reply:
> what are the minimum and maximum values that this identifier can conceivably be set to without underflowing or overflowing, and what types are most suitable for those limits in Objective-C and Swift?

32 bits. `UInt32` would be the more Obj-C choice here.

Reply:
`NSUInteger` and `UInt` are more ergonomic than `uint32_t` and `UInt32` in Objective-C and Swift, respectively (and the signed types are more ergonomic than the unsigned types). So as long as those more ergonomic types can fit all the possible values, then it’s OK even if the type isn’t exactly the same width.

Hypothetically, if a public read-write property uses a wider type than the underlying internal type, then the getter and application have nothing to worry about, but the setter needs some bounds checking. On the other hand, if the property uses the same type as the underlying code but it’s narrower than the more common type used in application code (such as `NSUInteger`), then the application ends up needing to bridge the impedance mismatch with explicit casts – that can be pretty error prone.

> Update: Interestingly it looks like `MGLAnnotationTag` is a `uint32_t` after all.

Ah, right. Perhaps of interest, `MGLAnnotationTagNotFound` is `UINT32_MAX`, under the assumption that `UINT32_MAX` − 1 is enough annotations for an uncontrived application.

https://github.com/mapbox/mapbox-gl-native/blob/44aa4f4be2bd504da982ffdcc0ffa5e84b2c6f31/platform/macos/src/MGLMapView.mm#L107